### PR TITLE
fix(platform-browser): configure `XhrFactory` to use `BrowserXhr`

### LINKED ISSUE #41311: ReferenceError: XMLHttpRequest is not defined
# 🐞 Bug report

#### What modules are related to this issue?

- [ ] aspnetcore-engine
- [x] builders
- [ ] common
- [x] express-engine
- [ ] hapi-engine

### Is this a regression?
IDK 😕😕

### Description

I am trying to make an http request via a pipe & I get the error below.

## 🔬 Minimal Reproduction
Here is a repo that show the bug: https://github.com/rabraghib/ng-universal-bug
the app setted up with the minimale possible changes using the CLI.
```shell
# generate new angular app
ng new ng-universal-bug --routing
# add angular universal
ng add @nguniversal/express-engine
# generate home module and route
ng g m home --route home --module app.module.ts
# generate http call pipe
ng g pipe home/httpCall
```
then I made some changes to let the pipe to make an http call to the given url .
& use it in home.component.html

## 🔥 Exception or Error
<pre><code>
ERROR ReferenceError: XMLHttpRequest is not defined
    at BrowserXhr.build (...\ng-universal-bug\dist\ng-universal-bug\server\main.js:150834:9)
    at Observable._subscribe (...\ng-universal-bug\dist\ng-universal-bug\server\main.js:150868:41)       
    at Observable._trySubscribe (...\ng-universal-bug\dist\ng-universal-bug\server\main.js:60530:25)     
    at Observable.subscribe (...\ng-universal-bug\dist\ng-universal-bug\server\main.js:60516:22)
    at innerSubscribe (...\ng-universal-bug\dist\ng-universal-bug\server\main.js:173928:23)
    at MergeMapSubscriber._innerSub (...\ng-universal-bug\dist\ng-universal-bug\server\main.js:42626:105)
    at MergeMapSubscriber._tryNext (...\ng-universal-bug\dist\ng-universal-bug\server\main.js:42620:14)  
    at MergeMapSubscriber._next (...\ng-universal-bug\dist\ng-universal-bug\server\main.js:42603:18)     
    at MergeMapSubscriber.next (...\ng-universal-bug\dist\ng-universal-bug\server\main.js:54386:18)      
    at Observable._subscribe (...\ng-universal-bug\dist\ng-universal-bug\server\main.js:131030:20) 
</code></pre>

## 🌍 Your Environment
<pre><code># ng version
Angular CLI: 11.2.5
Node: 12.18.0
OS: win32 x64

Angular: 11.2.6
... animations, common, compiler, compiler-cli, core, forms    
... platform-browser, platform-browser-dynamic, platform-server
... router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------      
@angular-devkit/architect       0.1102.5
@angular-devkit/build-angular   0.1102.5
@angular-devkit/core            11.2.5
@angular-devkit/schematics      11.2.5
@angular/cli                    11.2.5
@nguniversal/builders           11.2.1
@nguniversal/express-engine     11.2.1
@schematics/angular             11.2.5
@schematics/update              0.1102.5
rxjs                            6.6.6
typescript                      4.1.5
</code></pre>

### PR DESCRIPTION
**Pull request by alan-agius4**
**feat(core): add migration for  `XhrFactory` import**

Automatically migrates `XhrFactory` from `@angular/common/http` to `@angular/common`.

**build(docs-infra): support building the local `angular-in-memory-web-api` package in `NgPackagesInstaller`**

In some cases, we want to test the AIO app or docs examples against the locally built `angular-in-memory-web-api` for example to ensure that the changes in a commit do not introduce a breaking changes.

**fix(platform-browser): configure `XhrFactory` to use `BrowserXhr`**

With this change we move `XhrFactory` to the root entrypoint of `@angular/commmon`, this is needed so that we can configure `XhrFactory` DI token at a platform level, and not add a dependency between `@angular/platform-browser` and `@angular/common/http`.

Currently, when using `HttpClientModule` in a child module on the server, `ReferenceError: XMLHttpRequest is not defined` is being thrown because the child module has its own Injector and causes `XhrFactory` provider to be configured to use `BrowserXhr`.
Therefore, we should configure the `XhrFactory` at a platform level similar to other Browser specific providers.

**BREAKING CHANGE:**

`XhrFactory` has been moved from `@angular/common/http` to `@angular/common`.

**Before**
```ts
import {XhrFactory} from '@angular/common/http';
```

**After**
```ts
import {XhrFactory} from '@angular/common';
```

Closes #41311

### CODE REVIEW

File: packages/platform-browser/src/browser.ts [hunk 1]
```diff
@@ -66,7 +67,7 @@ export const BROWSER_SANITIZATION_PROVIDERS = BROWSER_SANITIZATION_PROVIDERS__PR
 export const platformBrowser: (extraProviders?: StaticProvider[]) => PlatformRef =
     createPlatformFactory(platformCore, 'browser', INTERNAL_BROWSER_PLATFORM_PROVIDERS);
 
-export const BROWSER_MODULE_PROVIDERS: StaticProvider[] = [
+export const BROWSER_MODULE_PROVIDERS: Provider[] = [
```
Comment (gkalpak) [inline packages/platform-browser/src/browser.ts:70]: Why this change? This doesn't seem desirable 😕

File: packages/core/schematics/migrations/xhr-factory/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,17 @@
+load("//tools:defaults.bzl", "ts_library")
+
+ts_library(
+    name = "xhr-factory",
+    srcs = glob(["**/*.ts"]),
+    tsconfig = "//packages/core/schematics:tsconfig.json",
+    visibility = [
+        "//packages/core/schematics:__pkg__",
+        "//packages/core/schematics/test:__pkg__",
```
Comment (gkalpak) [inline packages/core/schematics/migrations/xhr-factory/BUILD.bazel:9]: Shouldn't this be made visible to [//packages/core/schematics/migrations/google3:__pkg__][1] (as other migrations do)?
(DISCLAIMER: I have no idea what the `google3/` directory is for 🙊)

[1]: https://github.com/angular/angular/tree/78e6a9a5c11485ca21d9552aec26542646b562dc/packages/core/schematics/migrations/google3

File: packages/core/schematics/migrations/xhr-factory/index.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+              sourceFile);
+          recorder.remove(index, length).insertLeft(index, newImports);
+        } else {
+          // Remove @angular/common/http import
+          const index = httpCommonImport.getFullStart();
+          const length = httpCommonImport.getFullWidth();
+          recorder.remove(index, length);
+        }
+
+        // Import XhrFactory from @angular/comm on
```
Comment (gkalpak) [inline packages/core/schematics/migrations/xhr-factory/index.ts:89]:
```suggestion
        // Import XhrFactory from @angular/common
```

File: packages/core/schematics/migrations/xhr-factory/index.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        if (commonImport) {
+          const commonNamedBinding = commonImport.importClause?.namedBindings;
+          if (commonNamedBinding && ts.isNamedImports(commonNamedBinding)) {
+            // Already has an import for @angular/common, just add the named import.
+            const index = commonNamedBinding.getStart();
+            const length = commonNamedBinding.getWidth();
+            const newImports = printer.printNode(
+                ts.EmitHint.Unspecified,
+                ts.factory.updateNamedImports(
+                    namedBinding, [...commonNamedBinding.elements, xhrFactorySpecifier]),
```
Comment (gkalpak) [inline packages/core/schematics/migrations/xhr-factory/index.ts:100]: Shouldn't this be:

```suggestion
                    commonNamedBinding, [...commonNamedBinding.elements, xhrFactorySpecifier]),
```

File: packages/core/schematics/tsconfig.json [hunk 5]
```diff
@@ -3,7 +3,9 @@
     "noImplicitReturns": true,
     "noFallthroughCasesInSwitch": true,
     "strict": true,
-    "lib": ["es2015"],
+    "moduleResolution": "node",
+    "target": "es2019",
+    "lib": ["es2019"],
```
Comment (gkalpak) [inline packages/core/schematics/tsconfig.json:8]: Why these changes?
Also, how confident are we that they won't break anything? 🤔

File: packages/misc/angular-in-memory-web-api/package.json [hunk 6]
```diff
@@ -5,10 +5,12 @@
   "author": "angular",
   "license": "MIT",
   "peerDependencies": {
-    "@angular/core": "^8.0.0",
-    "@angular/common": "^8.0.0",
-    "rxjs": "^6.5.3",
-    "tslib": "^1.10.0"
+    "@angular/core": "^12.0.0-next.5",
+    "@angular/common": "^12.0.0-next.5",
```
Comment (gkalpak) [inline packages/misc/angular-in-memory-web-api/package.json:9]: OOC, why these changes?

File: scripts/build/angular-in-memory-web-api.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+'use strict';
+
+const {chmod, cp, mkdir, rm} = require('shelljs');
+const {baseDir, bazelBin, bazelCmd, exec, scriptPath} = require('./package-builder');
+
+/**
+ * Build the `angular-in-memory-web-api` npm package and copies it to `dist/packages-dist`.
```
Comment (gkalpak) [inline scripts/build/angular-in-memory-web-api.js:15]:
```suggestion
 * Build the `angular-in-memory-web-api` npm package and copy it to `dist/packages-dist/misc`.
```

**Review by gkalpak — CHANGES_REQUESTED** [review]
I'm not sure if this is the right fix (I'll let others comment on that), but reviewed for `dev-infra` and `docs-infra` :smiley:

File: packages/core/schematics/migrations/xhr-factory/BUILD.bazel (hunk 2 above)
Comment (alan-agius4) [reply packages/core/schematics/migrations/xhr-factory/BUILD.bazel:9]: We shouldn’t make targets visible unless necessary.

In this case it’s not needed, because there is no test for G3.

File: packages/core/schematics/tsconfig.json (hunk 5 above)
Comment (alan-agius4) [reply packages/core/schematics/tsconfig.json:8]: Schematics run on Node, and the lining supported version for v12, is Node.JS 12, which supports all ES2019 features.

This is needed because in the migration we use for...of with symbol iterators.

File: packages/misc/angular-in-memory-web-api/package.json (hunk 6 above)
Comment (alan-agius4) [reply packages/misc/angular-in-memory-web-api/package.json:9]: Because older versions will no longer be supported by this change. Also, the peer dep was wrong in the first place

File: packages/platform-browser/src/browser.ts (hunk 1 above)
Comment (alan-agius4) [reply packages/platform-browser/src/browser.ts:70]: It’s needed because BrowserXhr is not assignable to StaticProvider.

Comment (gkalpak) [reply packages/platform-browser/src/browser.ts:70]: Couldn't you just change it to `{provide: BrowserXhr, useClass: BrowserXhr, deps: []}`?

File: packages/core/schematics/migrations/xhr-factory/BUILD.bazel (hunk 2 above)
Comment (gkalpak) [reply packages/core/schematics/migrations/xhr-factory/BUILD.bazel:9]: > In this case it’s not needed, because there is no test for G3.

The question is: Should there be? 🤔

File: packages/misc/angular-in-memory-web-api/package.json (hunk 6 above)
Comment (gkalpak) [reply packages/misc/angular-in-memory-web-api/package.json:9]: But then shouldn't these be:

```suggestion
    "@angular/core": "^12.0.0-next.6",
    "@angular/common": "^12.0.0-next.6",
```

File: packages/platform-browser/src/browser.ts [hunk 8]
```diff
@@ -88,6 +89,7 @@ export const BROWSER_MODULE_PROVIDERS: StaticProvider[] = [
   {provide: DomSharedStylesHost, useClass: DomSharedStylesHost, deps: [DOCUMENT]},
   {provide: Testability, useClass: Testability, deps: [NgZone]},
   {provide: EventManager, useClass: EventManager, deps: [EVENT_MANAGER_PLUGINS, NgZone]},
+  {provide: XhrFactory, useClass: BrowserXhr},
```
Comment (gkalpak) [inline packages/platform-browser/src/browser.ts:92]: I believe this would allow us to continue using `StaticProvider[]`:

```suggestion
  {provide: XhrFactory, useClass: BrowserXhr, deps: []},
```

File: packages/core/schematics/migrations/xhr-factory/index.ts [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+      if (allImportDeclarations.length === 0) {
+        continue;
+      }
+
+      const httpCommonImport = findImportDeclaration('@angular/common/http', allImportDeclarations);
+      if (!httpCommonImport) {
+        continue;
+      }
+
+      const commonNamedBinding = httpCommonImport.importClause?.namedBindings;
```
Comment (gkalpak) [inline packages/core/schematics/migrations/xhr-factory/index.ts:59]: Calling this `commonNamedBinding` and having is shadowed below on line 93 makes the code difficult to follow.
I suggest naming this `commonHttpNamedBinding` (to indicate if refer to `@angular/common/http` vs just `@angular/common`).

File: packages/core/schematics/migrations/xhr-factory/BUILD.bazel (hunk 2 above)
Comment (alan-agius4) [reply packages/core/schematics/migrations/xhr-factory/BUILD.bazel:9]: Maybe @kyliau can check about the usage in G3?

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: packages/platform-browser/src/browser.ts (hunk 8 above)
Comment (alan-agius4) [reply packages/platform-browser/src/browser.ts:92]: Isn’t `deps: []` the same as when it’s omitted?

File: packages/core/schematics/migrations/xhr-factory/index.ts (hunk 9 above)
Comment (alan-agius4) [reply packages/core/schematics/migrations/xhr-factory/index.ts]: Its actually not shadowed, since they are in 2 different scopes, but I did rename it.

File: packages/platform-browser/src/browser.ts (hunk 8 above)
Comment (gkalpak) [reply packages/platform-browser/src/browser.ts:92]: For `StaticProvider` it is required:

https://github.com/angular/angular/blob/cbba7152fd2a533277b88aca98a50339d098d9f7/packages/core/src/di/interface/provider.ts#L66-L70

File: packages/core/schematics/migrations/xhr-factory/index.ts (hunk 9 above)
Comment (gkalpak) [reply packages/core/schematics/migrations/xhr-factory/index.ts]: <sub>I am pretty sure they were in the same scope 😛 - but good thing you renamed it anyway</sub>

**Review by gkalpak — APPROVED** [review]
LGTM (pending https://github.com/angular/angular/pull/41313#discussion_r599903171) :+1:

BTW, we need to ensure a new release for `angular-in-memory-web-api` is published (I don't think it is being published with the framework).

File: packages/core/schematics/test/xhr_factory_spec.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+      import { XhrFactory as XhrFactory2 } from '@angular/common/http';
+      import { Injecable } from '@angular/core';
+    `);
+
+    await runMigration();
+    expect(tree.readContent('/index.ts')).toBe(tags.stripIndents`
+      import { HttpClient, XhrFactory as XhrFactory2 } from '@angular/common';
+      import { Injecable } from '@angular/core';
+    `);
+  });
```
Comment (petebacondarwin) [inline packages/core/schematics/test/xhr_factory_spec.ts:61]: What about cases where there are namespace imports?

E.g.

```ts
import * as http from '@angular/common/http';
...
class MyXhrFactory extends http.XhrFactory { ... }
```

I don't think this first case is handled by this migration.

or 

```ts
import * as common from '@angular/common';
import {XhrFactory} from '@angular/common/http';
```

In this second case, will this migration just add a new statement?

```ts
import * as common from '@angular/common';
import {XhrFactory} from '@angular/common';
```

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Nice PR @alan-agius4!

I feel like these two commits

- fix(platform-browser): configure XhrFactory to use BrowserXhr
- feat(common): move XhrFactory from http entrypoint to root entrypoint

could be merged together and just be a single `fix` with a breaking change commit.

File: goldens/size-tracking/integration-payloads.json [hunk 11]
```diff
@@ -21,7 +21,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 1485,
-        "main-es2015": 144151,
+        "main-es2015": 144579,
```
[the remark this answers is not in the document]
Comment (alan-agius4) [reply goldens/size-tracking/integration-payloads.json:24]: The size checks here are uncompressed and hence without dead code elimination

File: packages/core/schematics/test/xhr_factory_spec.ts (hunk 10 above)
Comment (alan-agius4) [reply packages/core/schematics/test/xhr_factory_spec.ts:61]: Yeah it will add a new named import. Will add a test case.

Comment (alan-agius4) [reply packages/core/schematics/test/xhr_factory_spec.ts:61]: Re the first `import * as http from '@angular/common/http';` it is not handled. I am not sure if it's worth handling to be honest. Using `XhrFactory` is not that common, and using it with a namespace import is likely even less common. 

I think in this case, maybe we can add this following feedback.

File: packages/platform-browser/src/browser/xhr.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {XhrFactory} from '@angular/common';
+import {Injectable} from '@angular/core';
+
+/**
+ * A factory for `HttpXhrBackend` that uses the `XMLHttpRequest` browser API.
+ */
+@Injectable()
```
Comment (petebacondarwin) [inline packages/platform-browser/src/browser/xhr.ts:15]: I wonder if we should use a tree-shakable provider pattern here?

File: packages/core/schematics/test/xhr_factory_spec.ts (hunk 10 above)
Comment (petebacondarwin) [reply packages/core/schematics/test/xhr_factory_spec.ts:61]: I agree it is probably rare enough not to worry about.

**Review by petebacondarwin — APPROVED** [review]
LGTM - just waiting on feedback about G3?

https://github.com/angular/angular/pull/41313/commits/d08b1eb0c1c8445aea4e05dfbb2896dce9a81ca3#r600530942

File: packages/platform-browser/src/browser/xhr.ts (hunk 12 above)
Comment (alan-agius4) [reply packages/platform-browser/src/browser/xhr.ts:15]: I don’t think it will make any difference, since `BrowserXhr` is referenced in module providers as such

```ts
{provider: XhrFactory, useClass: BrowserXhr}
```

File: packages/core/schematics/migrations/xhr-factory/BUILD.bazel (hunk 2 above)
Comment (kyliau) [reply packages/core/schematics/migrations/xhr-factory/BUILD.bazel:9]: There are about 20 targets in google3 that uses `import {XhrFactory} from '@angular/common/http';`
I'll just do the migration manually - the internal CLI doesn't support `ng update`.
How should be roll out process be?
1. Re-export the symbol from the old location, 
2. sync into g3, 
3. manual migration,
4. Delete symbol from old location

File: packages/platform-browser/src/browser/xhr.ts (hunk 12 above)
Comment (petebacondarwin) [reply packages/platform-browser/src/browser/xhr.ts:15]: What I mean is that if we changed this to:

```ts
@Injectable({providedIn: BrowserModule})
```

Then I think we could drop the provider from the module itself, no?

Comment (petebacondarwin) [reply packages/platform-browser/src/browser/xhr.ts:15]: Not blocking landing this though!

Comment (alan-agius4) [reply packages/platform-browser/src/browser/xhr.ts:15]: Not really since we need to set `XhrFactory` provider to use `BrowserXhr`

i.e.

https://github.com/angular/angular/pull/41313/files#diff-37c00887df258c330c3953e9349bec006b6374b8586431184164d650e122daa7R92

Cannot see how the tree-shakable provider replaces this.

File: packages/core/schematics/migrations/xhr-factory/BUILD.bazel (hunk 2 above)
Comment (alan-agius4) [reply packages/core/schematics/migrations/xhr-factory/BUILD.bazel:9]: I think this makes sense, if we remove the re-export  before v12 is final.

The reason for this is that there is no good way to `@deprecate` the re-export without breaking DI.

File: packages/platform-browser/src/browser/xhr.ts (hunk 12 above)
Comment (petebacondarwin) [reply packages/platform-browser/src/browser/xhr.ts:15]: Doh! Of course. Nevermind me 😸

File: packages/core/schematics/migrations/xhr-factory/BUILD.bazel (hunk 2 above)
Comment (kyliau) [reply packages/core/schematics/migrations/xhr-factory/BUILD.bazel:9]: yeah since it's only 20 targets, I'll just do it manually.

Comment (alan-agius4) [reply packages/core/schematics/migrations/xhr-factory/BUILD.bazel:9]: Re-export added.

**Review by kyliau — APPROVED** [review]

File: packages/core/schematics/migrations.json [hunk 13]
```diff
@@ -89,6 +89,11 @@
       "version": "12.0.0-beta",
       "description": "In Angular version 12, the type of ActivatedRouteSnapshot.fragment is nullable. This migration automatically adds non-null assertions to it.",
       "factory": "./migrations/activated-route-snapshot-fragment/index"
+    },
+    "migration-v12-xhr-factory": {
+      "version": "12.0.0-next.6",
+      "description": "`XhrFactory` migration. `XhrFactory` has been moved from `@angular/common/http` to `@angular/common`.",
```
Comment (crisbeto) [inline packages/core/schematics/migrations.json:95]: I think that you can drop the "`XhrFactory` migration." since it's inferred by the migration name already.

File: packages/core/schematics/migrations/xhr-factory/index.ts [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+
+    yield* visit(directory.dir(path));
+  }
+}
+
+export default function(): Rule {
+  return tree => {
+    const printer = ts.createPrinter({newLine: ts.NewLineKind.LineFeed});
+
+    for (const sourceFile of visit(tree.root)) {
```
Comment (crisbeto) [inline packages/core/schematics/migrations/xhr-factory/index.ts:45]: This whole block is somewhat long. Can it be broken up into functions? E.g. the logic for finding the imports can be moved out into a helper.

**Review by crisbeto — APPROVED** [review]
LGTM with a couple of nits.

**Review by mhevery — APPROVED** [review]

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

**Review by jelbourn — APPROVED** [review]
LGTM

Reviewed-for: public-api

**Review by petebacondarwin — APPROVED** [review]
Reviewed-for: public-api

**Review by petebacondarwin — APPROVED** [review]
Reviewed-for: size-tracking

**Review by atscott — APPROVED** [review]
Reviewed-for: size-tracking

**Comment by kyliau** [comment]
g3 presubmit:	http://test/OCL:365647464:BASE:365647516:1617048694291:251772e6
